src/async_reader.rs

```diff
@@ -188,17 +193,32 @@ impl AsyncPmTilesReader<MmapBackend> {
 
 #[async_trait]
 pub trait AsyncBackend {
-    async fn read_bytes(&self, dst: &mut [u8], offset: usize) -> Result<(), Error>;
+    /// Reads exactly `dst.len()` starting at [offset]
+    async fn read_exact(&self, dst: &mut [u8], offset: usize) -> Result<(), Error>;
+
+    /// Reads up to `dst.len()` bytes into [dst].
+    async fn read(&self, dst: &mut [u8], offset: usize) -> Result<usize, Error>;
+
+    /// Read the first 127 and up to 16,384 bytes to ensure we can initialize the header and root directory.
+    async fn read_initial_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut bytes = vec![0; MAX_INITIAL_BYTES];
+        let n_read_bytes = self.read(bytes.as_mut_slice(), 0).await?;
+        if n_read_bytes < HEADER_SIZE {
+            return Err(Error::InvalidHeader);
+        }
+
+        Ok(bytes)
+    }
 
     async fn read_header_bytes(&self) -> Result<[u8; 127], Error> {
         let mut header_bytes = [0; 127];
-        self.read_bytes(&mut header_bytes, 0).await?;
+        self.read_exact(&mut header_bytes, 0).await?;
 
         Ok(header_bytes)
     }
 }
 
-#[cfg(test)]
+#[cfg(all(feature = "mmap-async-tokio", test))]
```

Comment:
This feels wrong,  tests should not depend on some feature being enabled. Unit tests should just assume every feature is on. I could be wrong

Reply:
Oops. Good catch. I thought I had removed them all, after trying a few different approaches. 😄